framework/src/media/BufferOutputDataSource.cpp

```diff
@@ -57,57 +61,47 @@ BufferOutputDataSource &BufferOutputDataSource::operator=(const BufferOutputData
 
 bool BufferOutputDataSource::open()
 {
-	if (!getStreamBuffer()) {
-		auto streamBuffer = StreamBuffer::Builder()
-								.setBufferSize(CONFIG_BUFFER_DATASOURCE_STREAM_BUFFER_SIZE)
-								.setThreshold(CONFIG_BUFFER_DATASOURCE_STREAM_BUFFER_THRESHOLD)
-								.build();
-		if (!streamBuffer) {
-			medvdbg("streamBuffer is nullptr!\n");
-			return false;
-		}
-
-		setStreamBuffer(streamBuffer);
-	}
-
-	start();
+	mIsPrepare = true;
 	return true;
 }
 
 bool BufferOutputDataSource::close()
 {
-	stop();
+	mIsPrepare = false;
 	return true;
 }
 
 bool BufferOutputDataSource::isPrepare()
 {
-	return (getStreamBuffer() != nullptr);
+	return mIsPrepare;
 }
 
-ssize_t BufferOutputDataSource::onStreamBufferReadable(bool isFlush)
+ssize_t BufferOutputDataSource::write(unsigned char *buf, size_t size)
 {
-	auto reader = getBufferReader();
-	assert(reader->sizeOfData() > 0);
-	size_t size = reader->sizeOfData();
-	unsigned char *buffer = new unsigned char[size];
-	if (buffer) {
-		reader->read(buffer, size);
-		auto recorder = getRecorder();
-		if (recorder) {
-			recorder->notifyObserver(OBSERVER_COMMAND_BUFFER_DATAREACHED, buffer, size);
-		}
-		// DO NOT `delete[]`, `buffer` will be managed in std::shared_ptr<> and released automatically!
-		return size;
+	if (!isPrepare()) {
+		return EOF;
+	}
+
+	if (buf == nullptr) {
+		return EOF;
 	}
 
-	meddbg("buffer alloc failed\n");
-	return -1;
+	auto buffer = new unsigned char[size];
+	memcpy(buffer, buf, size);
+	auto recorder = getRecorder();
+	if (recorder) {
+		recorder->notifyObserver(OBSERVER_COMMAND_BUFFER_DATAREACHED, buffer, size);
+	}
+	// DO NOT `delete[]`, `buffer` will be managed in std::shared_ptr<> and released automatically!
+
+	return size;
 }
 
 BufferOutputDataSource::~BufferOutputDataSource()
 {
-	close();
+	if (isPrepare()) {
```

Comment:
If isPrepare is checking in Class, is it better to 'mIsPrepare'?

Reply:
All dataSource destructor can have same code, it is better for readability